cmd/scollector/collectors/snmp_ciscobgp.go

```diff
@@ -0,0 +1,190 @@
+		if err != nil {
+			return err
+		}
+		for k, v := range m {
+			tags := opentsdb.TagSet{
+				"host":      host,
+				"peer":      k,
+				"local_as":  localASes[k],
+				"local_id":  localIdentifiers[k],
+				"remote_as": remoteASes[k],
+				"remote_id": remoteIdentifiers[k],
+			}
+			Add(&md, bA.metric, v, tags, bA.rate, bA.unit, bA.desc)
+		}
+		return nil
+	}
+	// oid, metric, rate, unit, description
+	oids := []bgpAdd{
+		{state, bgpPeerState, metadata.Gauge, metadata.StatusCode, bgpPeerStateDesc},
+		{adminStatus, bgpPeerAdminStatus, metadata.Gauge, metadata.StatusCode, bgpPeerAdminStatusDesc},
+		{inUpdates, bgpPeerInUpdates, metadata.Counter, metadata.Packet, bgpPeerInUpdatesDesc},
+		{outUpdates, bgpPeerOutUpdates, metadata.Counter, metadata.Packet, bgpPeerOutUpdatesDesc},
+		{inTotalMessages, bgpPeerInTotalMessages, metadata.Counter, metadata.Packet, bgpPeerInTotalMessagesDesc},
+		{outTotalMessages, bgpPeerOutTotalMessages, metadata.Counter, metadata.Packet, bgpPeerOutTotalMessagesDesc},
+		{inUpdateElapsedTime, bgpPeerInUpdateElapsedTime, metadata.Gauge, metadata.Second, bgpPeerInUpdateElapsedTimeDesc},
+		{acceptedPrefixes, bgpPeerAcceptedPrefixes, metadata.Gauge, metadata.Count, bgpPeerAcceptedPrefixesDesc},
+		{deniedPrefixes, bgpPeerDeniedPrefixes, metadata.Counter, metadata.Count, bgpPeerDeniedPrefixesDesc},
+		{advertisedPrefixes, bgpPeerAdvertisedPrefixes, metadata.Counter, metadata.Count, bgpPeerAdvertisedPrefixesDesc},
+		{suppressedPrefixes, bgpPeerSuppressedPrefixes, metadata.Counter, metadata.Count, bgpPeerSuppressedPrefixesDesc},
+		{withdrawnPrefixes, bgpPeerWithdrawnPrefixes, metadata.Counter, metadata.Count, bgpPeerWithdrawnPrefixesDesc},
+	}
+	for _, bA := range oids {
+		if err := add(bA); err != nil {
+			return nil, err
+		}
+	}
+	return md, nil
+}
+
+func snmp_ip_tree(host, community, oid string) (map[string]interface{}, error) {
+	rows, err := snmp.Walk(host, community, oid)
+	if err != nil {
+		return nil, err
+	}
+	m := make(map[string]interface{})
+	for rows.Next() {
+		key := ""
+		var a interface{}
+		id, err := rows.Scan(&a)
+		if err != nil {
+			return nil, err
+		}
+		switch t := id.(type) {
+		case []int:
+			key = snmp_combine_ip_int(t)
+		default:
+			return nil, fmt.Errorf("wrong type?")
+		}
+		m[key] = a
+	}
+	if err := rows.Err(); err != nil && err != io.EOF {
+		return nil, err
+	}
+	return m, nil
+}
+
+func snmp_combine_ip_uint8(path []uint8) string {
+	s := make([]string, len(path))
+	for i := range path {
+		s[i] = fmt.Sprint(path[i])
+	}
+	// In some cases there's a .1.1 after the IP (because SNMP is awesome), drop it:
+	s = s[:4]
+	return strings.Join(s, ".")
+}
+
+func snmp_combine_ip_int(path []int) string {
+	s := make([]string, len(path))
+	for i := range path {
+		s[i] = fmt.Sprint(path[i])
+	}
+	// In some cases there's a .1.1 after the IP (because SNMP is awesome), drop it:
+	s = s[:4]
+	return strings.Join(s, ".")
+}
+
+type bgpAdd struct {
```

Comment:
this struct could even be defined inside of your collector function. I am getting pretty serious about not wanting to add uneccessary things to the collectors package namespace.